lib/Sema/CSSyntacticElement.cpp

```diff
@@ -1023,14 +1023,23 @@ class SyntacticElementConstraintGenerator
 
     SmallVector<ElementInfo, 4> elements;
 
-    // First, let's record a body of `do` statement.
-    elements.push_back(makeElement(doStmt->getBody(), doLoc));
+    // First, let's record a body of `do` statement. Note we need to add a
+    // SyntaticElement locator path element here to avoid treating the inner
+    // brace conjunction as being isolated if 'doLoc' is for an isolated
+    // conjunction (as is the case with 'do' expressions).
+    auto *doBodyLoc = cs.getConstraintLocator(
+        doLoc, LocatorPathElt::SyntacticElement(doStmt->getBody()));
+    elements.push_back(makeElement(doStmt->getBody(), doBodyLoc));
 
     // After that has been type-checked, let's switch to
     // individual `catch` statements.
     for (auto *catchStmt : doStmt->getCatches())
       elements.push_back(makeElement(catchStmt, doLoc));
 
+    // Inject a join if we have one.
+    if (auto *join = context.ElementJoin.getPtrOrNull())
```

Comment:
Wouldn't it be safer to perform this operation do it in `createConjunction` that way it could be added directly to the new constraint too?...

Reply:
Heh yeah, I was hoping to refactor this further by having SyntacticElementConstraintGenerator store the elements and have a single call to build the conjunction, but yeah this is a good intermediate step, how does it look?

Reply:
Looks good but I think you forgot to remove all the existing inline uses

Reply:
Looks good, thank you!